File: Java/261. Graph Valid Tree.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        DSU dsu = new DSU(n);
        
        for(int[] edge : edges){
            int u = edge[0];
            int v = edge[1];
            if(dsu.connected(u, v)){
                return false;
            }
            dsu.union(u, v);
        }
        
        return dsu.isTree();
    }
}

class DSU{
    private int[] root;
    private int[] rank;
    private int count;
    
    DSU(int size){
        root = new int[size];
        rank = new int[size];
        count = size;
        
        for(int i=0; i<size; i++){
            root[i] = i;
            rank[i] = 1;
        }
    }
    
    public int find(int x){
        return x == root[x] ? x : (root[x] = find(root[x]));
    }
    
    public void union(int a, int b){
        int x = root[a];
        int y = root[b];
        
        if(x != y){
            count--;
            if(rank[x] < rank[y]){
                root[x] = y;
            }else if(rank[x] > rank[y]){
                root[y] = x;
            }else{
                root[x] = y;
                rank[y]++;
            }
        }
    }
    
    public boolean connected(int x, int y){
        return find(x) == find(y);
    }
    
    public boolean isTree(){
        return count == 1;
    }
}

File: Python3/261. Graph Valid Tree.py
class DSU:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        self.rank = [1]*size
        self.count = size
        
    def find(self, x):
        if x == self.root[x]:
            return x
        self.root[x] = self.find(self.root[x])
        return self.root[x]
        
    def union(self, a, b):
        x = self.root[a]
        y = self.root[b]
        
        if x != y:
            self.count -= 1
            if self.rank[x] < self.rank[y]:
                self.root[x] = y
            elif self.rank[x] > self.rank[y]:
                self.root[y] = x
            else:
                self.root[x] = y
                self.rank[y] += 1
    
    # if two nodes are already connected
    # adding an edge between them will introduce a cycle
    def connected(self, x, y):
        return self.find(x) == self.find(y)
    
    # if self.count == 1, it means there's only one root for all the nodes
    def singleRoot(self):
        return True if self.count == 1 else False
    

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        dsu = DSU(n)
        
        for edge in edges:
            u, v = edge[0], edge[1]
            if dsu.connected(u, v):
                return False
            
            # merge the two nodes
            dsu.union(u, v)
            
        return dsu.singleRoot()